File: nlox/Lox/Resolver.cs
class Resolver
{
  private Interpreter interpreter;
  private List<Dictionary<string, bool>> scopes = new();
  private FunctionType currentFunction = FunctionType.None;
  private ClassType currentClass = ClassType.None;

  public Resolver(Interpreter interpreter)
  {
    this.interpreter = interpreter;
  }

  public void Resolve(List<Stmt> statements)
  {
    foreach (var stmt in statements)
    {
      Resolve(stmt);
    }
  }

  private void Resolve(Stmt stmt)
  {
    switch (stmt)
    {
      case Block block:
        Resolve(block);
        break;
      case ClassDecl @class:
        Resolve(@class);
        break;
      case Var var:
        Resolve(var);
        break;
      case Function function:
        Resolve(function);
        break;
      case ExpressionStmt exprStmt:
        Resolve(exprStmt);
        break;
      case If @if:
        Resolve(@if);
        break;
      case Print print:
        Resolve(print);
        break;
      case Return @return:
        Resolve(@return);
        break;
      case While @while:
        Resolve(@while);
        break;
      default:
        throw new Exception("Unknown statement");
    }
  }

  private void Resolve(Block stmt)
  {
    BeginScope();
    Resolve(stmt.Statements);
    EndScope();
  }

  private void Resolve(ClassDecl stmt)
  {
    var enclosingClass = currentClass;
    currentClass = ClassType.Class;

    Declare(stmt.Name);
    Define(stmt.Name);

    if (stmt.Superclass != null && stmt.Name.Lexeme == stmt.Superclass.Name.Lexeme)
    {
      Lox.Error(stmt.Superclass.Name, "A class can't inherit from itself.");
    }
    
    if (stmt.Superclass != null)
    {
      currentClass = ClassType.Subclass;
      Resolve(stmt.Superclass);
    }

    if (stmt.Superclass != null)
    {
      BeginScope();
      scopes.Last()["super"] = true;
    }

    BeginScope();
    scopes.Last()["this"] = true;

    foreach (var method in stmt.Methods)
    {
      var declaration = FunctionType.Method;
      if (method.Name.Lexeme == "init")
      {
        declaration = FunctionType.Initializer;
      }

      ResolveFunction(method, declaration);
    }

    EndScope();

    if (stmt.Superclass != null)
    {
      EndScope();
    }

    currentClass = enclosingClass;
  }

  private void Resolve(Var stmt)
  {
    Declare(stmt.Name);
    if (stmt.Initializer != null)
    {
      Resolve(stmt.Initializer);
    }
    Define(stmt.Name);
  }

  private void Resolve(Function stmt)
  {
    Declare(stmt.Name);
    Define(stmt.Name);
    ResolveFunction(stmt, FunctionType.Function);
  }

  private void ResolveFunction(Function function, FunctionType type)
  {
    var enclosingFunction = currentFunction;
    currentFunction = type;

    BeginScope();
    foreach (var param in function.Params)
    {
      Declare(param);
      Define(param);
    }
    Resolve(function.Body);
    EndScope();

    currentFunction = enclosingFunction;
  }

  private void Resolve(ExpressionStmt stmt)
  {
    Resolve(stmt.Expression);
  }

  private void Resolve(If stmt)
  {
    Resolve(stmt.Condition);
    Resolve(stmt.ThenBranch);
    if (stmt.ElseBranch != null)
    {
      Resolve(stmt.ElseBranch);
    }
  }

  private void Resolve(Print stmt)
  {
    Resolve(stmt.Expression);
  }

  private void Resolve(Return stmt)
  {
    if (currentFunction == FunctionType.None)
    {
      Lox.Error(stmt.Keyword, "Can't return from top-level code.");
    }

    if (stmt.Value != null)
    {
      if (currentFunction == FunctionType.Initializer)
      {
        Lox.Error(stmt.Keyword, "Can't return a value from an initializer.");
      }

      Resolve(stmt.Value);
    }
  }

  private void Resolve(While stmt)
  {
    Resolve(stmt.Condition);
    Resolve(stmt.Body);
  }

  private void Resolve(Expr expr)
  {
    switch (expr)
    {
      case Variable variable:
        Resolve(variable);
        break;
      case Assign assign:
        Resolve(assign);
        break;
      case Binary binary:
        Resolve(binary);
        break;
      case Call call:
        Resolve(call);
        break;
      case Grouping grouping:
        Resolve(grouping);
        break;
      case Literal _:
        break;
      case Logical logical:
        Resolve(logical);
        break;
      case Unary unary:
        Resolve(unary);
        break;
      case Get get:
        Resolve(get);
        break;
      case Set set:
        Resolve(set);
        break;
      case This @this:
        Resolve(@this);
        break;
      case Super super:
        Resolve(super);
        break;
      default:
        throw new Exception("Unknown expression");
    }
  }

  private void Resolve(Variable expr)
  {
    if (scopes.Count > 0 && scopes.Last().TryGetValue(expr.Name.Lexeme, out bool defined) && !defined)
    {
      Lox.Error(expr.Name, "Can't read local variable in its own initializer");
    }

    ResolveLocal(expr, expr.Name);
  }

  private void Resolve(Assign expr)
  {
    Resolve(expr.Value);
    ResolveLocal(expr, expr.Name);
  }

  private void Resolve(Binary expr)
  {
    Resolve(expr.Left);
    Resolve(expr.Right);
  }

  private void Resolve(Call expr)
  {
    Resolve(expr.Callee);
    foreach (var argument in expr.Arguments)
    {
      Resolve(argument);
    }
  }

  private void Resolve(Grouping expr)
  {
    Resolve(expr.Expression);
  }

  private void Resolve(Logical expr)
  {
    Resolve(expr.Left);
    Resolve(expr.Right);
  }

  private void Resolve(Unary expr)
  {
    Resolve(expr.Right);
  }

  private void Resolve(Get expr)
  {
    Resolve(expr.Object);
  }

  private void Resolve(Set expr)
  {
    Resolve(expr.Value);
    Resolve(expr.Object);
  }

  private void Resolve(This expr)
  {
    if (currentClass == ClassType.None)
    {
      Lox.Error(expr.Keyword, "Can't use 'this' outside of a class.");
    }

    ResolveLocal(expr, expr.Keyword);
  }

  private void Resolve(Super expr)
  {
    if (currentClass == ClassType.None)
    {
      Lox.Error(expr.Keyword, "Can't use 'super' outside of a class.");
    }
    else if (currentClass == ClassType.Class)
    {
      Lox.Error(expr.Keyword, "Can't use 'super' in a class with no superclass.");
    }

    ResolveLocal(expr, expr.Keyword);
  }

  private void ResolveLocal(Expr expr, Token name)
  {
    for (int i = scopes.Count - 1; i >= 0; i--)
    {
      if (scopes[i].ContainsKey(name.Lexeme))
      {
        interpreter.Resolve(expr, scopes.Count - 1 - i);
      }
    }
  }

  private void BeginScope()
  {
    scopes.Add(new Dictionary<string, bool>());
  }

  private void EndScope()
  {
    scopes.RemoveAt(scopes.Count - 1);
  }

  private void Declare(Token name)
  {
    if (scopes.Count == 0)
    {
      return;
    }
    var scope = scopes.Last();
    if (scope.ContainsKey(name.Lexeme))
    {
      Lox.Error(name, "There already is a variable with this name in this scope.");
    }
    
    scope[name.Lexeme] = false;
  }

  private void Define(Token name)
  {
    if (scopes.Count == 0)
    {
      return;
    }
    scopes.Last()[name.Lexeme] = true;
  }
}

File: tslox/src/resolver.ts
import { ClassType } from "./classtype";
import { error } from "./error";
import { Assign, Binary, Call, Expr, Get, Grouping, Logical, Set, Super, This, Unary, Variable } from "./expr";
import { FunctionType } from "./functiontype";
import { Block, ClassDecl, ExpressionStmt, Function, If, Print, Return, Stmt, Var, While } from "./stmt";
import { Token } from "./token";

export type ResolveLocal = (expr: Expr, depth: number) => void;

type Scope = Map<string, boolean>;

export class Resolver {
  private scopes: Scope[] = [];
  private currentFunction: FunctionType = "None";
  private currentClass: ClassType = "None";

  constructor(private resolveLocalFn: ResolveLocal) { }

  public resolve(statements: Stmt[]) {
    for (const stmt of statements) {
      this.resolveStmt(stmt);
    }
  }

  private resolveStmt(stmt: Stmt) {
    switch (stmt.kind) {
      case "Block":
        this.resolveBlock(stmt);
        break;
      case "ClassDecl":
        this.resolveClass(stmt);
        break;
      case "Var":
        this.resolveVar(stmt);
        break;
      case "If":
        this.resolveIf(stmt);
        break;
      case "Print":
        this.resolvePrint(stmt);
        break;
      case "Return":
        this.resolveReturn(stmt);
        break;
      case "While":
        this.resolveWhile(stmt);
        break;
      case "Function":
        this.resolveFunction(stmt);
        break;
      case "ExpressionStmt":
        this.resolveExpressionStmt(stmt);
        break;
      default:
        throw new Error("Unknow statement");
    }
  }
  
  private resolveBlock(stmt: Block) {
    this.beginScope();
    this.resolve(stmt.statements);
    this.endScope();
  }

  private resolveClass(stmt: ClassDecl) {
    const enclosingClass = this.currentClass;
    this.currentClass = "Class";

    this.declare(stmt.name);
    this.define(stmt.name);

    if (stmt.superclass && stmt.name.lexeme === stmt.superclass.name.lexeme) {
      error(stmt.superclass.name, "A class can't inherit from itself.");
    }

    if (stmt.superclass) {
      this.currentClass = "Subclass";
      this.resolveVariable(stmt.superclass);
    }

    if (stmt.superclass) {
      this.beginScope();
      const scope = this.scopes[this.scopes.length - 1];
      scope.set("super", true);
    }

    this.beginScope();
    const scope = this.scopes[this.scopes.length - 1];
    scope.set("this", true);

    for (const method of stmt.methods) {
      let functionType: FunctionType = "Method";
      if (method.name.lexeme === "init") {
        functionType = "Initializer";
      }

      this.resolveAnyFunction(method, functionType);
    }

    this.endScope();

    if (stmt.superclass) {
      this.endScope();
    }

    this.currentClass = enclosingClass;
  }

  private resolveIf(stmt: If) {
    this.resolveExpr(stmt.condition);
    this.resolveStmt(stmt.thenBranch);

    if (stmt.elseBranch) {
      this.resolveStmt(stmt.elseBranch);
    }
  }

  private resolvePrint(stmt: Print) {
    this.resolveExpr(stmt.expression);
  }

  private resolveFunction(stmt: Function) {
    this.declare(stmt.name);
    this.define(stmt.name);
    this.resolveAnyFunction(stmt, "Function");
  }

  private resolveAnyFunction(stmt: Function, functionType: FunctionType) {
    const enclosingFunction = this.currentFunction;
    this.currentFunction = functionType;

    this.beginScope();
    for (const param of stmt.params) {
      this.declare(param);
      this.define(param);
    }
    this.resolve(stmt.body);
    this.endScope();

    this.currentFunction = enclosingFunction;
  }

  private resolveWhile(stmt: While) {
    this.resolveExpr(stmt.condition);
    this.resolveStmt(stmt.body);
  }

  private resolveReturn(stmt: Return) {
    if (this.currentFunction === "None") {
      error(stmt.keyword, "Can't return from top-level code.");
    }

    if (stmt.value) {
      if (this.currentFunction === "Initializer") {
        error(stmt.keyword, "Can't return from an initializer.");
      }

      this.resolveExpr(stmt.value);
    }
  }

  private resolveVar(stmt: Var) {
    this.declare(stmt.name);
    if (stmt.initializer) {
      this.resolveExpr(stmt.initializer);
    }
    this.define(stmt.name);
  }

  private resolveExpressionStmt(stmt: ExpressionStmt) {
    this.resolveExpr(stmt.expression);
  }

  private resolveExpr(expr: Expr) {
    switch (expr.kind) {
      case "Literal":
        break;
      case "Grouping":
        this.resolveGrouping(expr);
        break;
      case "Variable":
        this.resolveVariable(expr);
        break;
      case "This":
        this.resolveThis(expr);
        break;
      case "Super":
        this.resolveSuper(expr);
        break;
      case "Call":
        this.resolveCall(expr);
        break;
      case "Get":
        this.resolveGet(expr);
        break;
      case "Set":
        this.resolveSet(expr);
        break;
      case "Unary":
        this.resolveUnary(expr);
        break;
      case "Binary":
        this.resolveBinary(expr);
        break;
      case "Logical":
        this.resolveLogical(expr);
        break;
      case "Assign":
        this.resolveAssign(expr);
        break;
      default:
        throw new Error("Unknown expression");
    }
  }

  private resolveGrouping(expr: Grouping) {
    this.resolveExpr(expr.expression);
  }

  private resolveVariable(expr: Variable) {
    if (this.scopes.length > 0) {
      const scope = this.scopes[this.scopes.length - 1];
      if (scope.has(expr.name.lexeme)) {
        if (scope.get(expr.name.lexeme) === false) {
          error(expr.name, "Can't read local variable in its own initializer");
        }
      }
    }

    this.resolveLocal(expr, expr.name);
  }

  private resolveThis(expr: This) {
    if (this.currentClass === "None") {
      error(expr.keyword, "Can't use 'this' outside of a class.");
    }

    this.resolveLocal(expr, expr.keyword);
  }

  private resolveSuper(expr: Super) {
    if (this.currentClass === "None") {
      error(expr.keyword, "Can't use 'super' outside of a class.");
    } else if (this.currentClass === "Class") {
      error(expr.keyword, "Can't use 'super' in a class no superclass.");
    }

    this.resolveLocal(expr, expr.keyword);
  }

  private resolveCall(expr: Call) {
    this.resolveExpr(expr.callee);
    for (const arg of expr.arguments) {
      this.resolveExpr(arg);
    }
  }

  private resolveGet(expr: Get) {
    this.resolveExpr(expr.object);
  }

  private resolveSet(expr: Set) {
    this.resolveExpr(expr.value);
    this.resolveExpr(expr.object);
  }

  private resolveUnary(expr: Unary) {
    this.resolveExpr(expr.right);
  }

  private resolveBinary(expr: Binary) {
    this.resolveExpr(expr.left);
    this.resolveExpr(expr.right);
  }

  private resolveLogical(expr: Logical) {
    this.resolveExpr(expr.left);
    this.resolveExpr(expr.right);
  }

  private resolveAssign(expr: Assign) {
    this.resolveExpr(expr.right);
    this.resolveLocal(expr, expr.name);
  }
  
  private resolveLocal(expr: Expr, name: Token) {
    for (let i = this.scopes.length - 1; i >= 0; i--) {
      if (this.scopes[i].has(name.lexeme)) {
        this.resolveLocalFn(expr, this.scopes.length - 1 - i);
        return;
      }
    }
  }

  private beginScope() {
    this.scopes.push(new Map<string, boolean>());
  }

  private endScope() {
    this.scopes.pop();
  }

  private declare(name: Token) {
    if (this.scopes.length === 0) {
      return;
    }
    const scope = this.scopes[this.scopes.length - 1];

    if (scope.has(name.lexeme)) {
      error(name, "There already is a variable with this name in this scope.");
    }

    scope.set(name.lexeme, false);
  }

  private define(name: Token) {
    if (this.scopes.length === 0) {
      return;
    }
    const scope = this.scopes[this.scopes.length - 1];
    scope.set(name.lexeme, true);
  }
}
